![Housing In India](https://github.com/mobadara/housing-in-india/blob/main/assets/images/houses.jpeg?raw=1)

# 🏠 **Housing Prices in India: Exploratory Data Analysis**


[![Open in Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/mobadara/housing-in-india/blob/main/notebooks/exploratory-data-analysis.ipynb)


## 🧭 **Project Overview**

India’s housing market is diverse and dynamic, shaped by rapid urbanization, regional disparities, and evolving economic factors. Understanding property price variations across different regions requires more than just statistical summaries — it calls for an analytical deep dive into **location, infrastructure, and socioeconomic variables**.

This notebook focuses on performing an **Exploratory Data Analysis (EDA)** to understand the structure and relationships within the dataset before any cleaning or modeling is performed.  
Insights gained from this analysis will directly guide **data cleaning, transformation, and feature engineering** in the next phase.

---

## 🗂️ **Notebook Outline**
0. Setup
1. Dataset Inspection
2. Descriptive Statistics
3. Distribution Analysis  
4. Feature Relationships  
5. Outlier Detection  
6. Missing Data Patterns  
7. Cleaning & Transformation Recommendations  

---


> 💡 **Note:** This notebook focuses purely on analysis.  
> All cleaning, preprocessing, and modeling steps will be implemented in subsequent notebooks for reproducibility and clarity.

---

## **Setup**
In this phase, I use imported the required python packages (local and those installed using `pip`), downloaded the dataset using `kagglehub` package and put the dataset in the right format. The downloaded dataset contains three `csv` files: `train.csv`, `test.csv` and `sample_submission.csv`. **Sample Submission** is a **one-column** dataset, column being the predicted **house price** from the **test dataset** (`test.csv`). The test dataset does not contain the target (price in lacs).

The version of the packages used are also printed for reproducibility and reporting purposes.

In [1]:
# Import necessary packages
import kagglehub
import os
import sys
import pandas as pd

In [2]:
# Download Dataset
# Download latest version
# try:
#     path = kagglehub.dataset_download("anmolkumar/house-price-prediction-challenge")
# except Exception as e:
#     print(f"Error downloading dataset: {e}")
#     print("Use the available dataset in the data folder.")

In [3]:
# Set up Environment
MODULE_PATH = os.path.abspath(os.path.join("..", "src"))
if not os.path.exists("../data"):
    os.makedirs('../data')
if MODULE_PATH not in sys.path:
    sys.path.insert(0, MODULE_PATH)
!cp -r {path}/* ../data/
!rm -r {path}
print(f'Dataset downloaded to ../data/')
print(f'Files in ../data/: {os.listdir("../data/")}')

cp: cannot stat '{path}/*': No such file or directory
rm: cannot remove '{path}': No such file or directory
Dataset downloaded to ../data/
Files in ../data/: ['sample_submission.csv', 'train.csv', 'test.csv']


In [4]:
# Import personal packages
from analysis.inspection import Inspector

## **Dataset Inspection**

In [5]:
df_train = pd.read_csv("../data/train.csv");

In [6]:
inspector = Inspector(data=df_train)
inspector.inspect()

Dataset Shape: (29451, 12)

Data Types:
------------------------------


POSTED_BY                 object
UNDER_CONSTRUCTION         int64
RERA                       int64
BHK_NO.                    int64
BHK_OR_RK                 object
SQUARE_FT                float64
READY_TO_MOVE              int64
RESALE                     int64
ADDRESS                   object
LONGITUDE                float64
LATITUDE                 float64
TARGET(PRICE_IN_LACS)    float64
dtype: object


Basic Information:
------------------------------
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 29451 entries, 0 to 29450
Data columns (total 12 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   POSTED_BY              29451 non-null  object 
 1   UNDER_CONSTRUCTION     29451 non-null  int64  
 2   RERA                   29451 non-null  int64  
 3   BHK_NO.                29451 non-null  int64  
 4   BHK_OR_RK              29451 non-null  object 
 5   SQUARE_FT              29451 non-null  float64
 6   READY_TO_MOVE          29451 non-null  int64  
 7   RESALE                 29451 non-null  int64  
 8   ADDRESS                29451 non-null  object 
 9   LONGITUDE              29451 non-null  float64
 10  LATITUDE               29451 non-null  float64
 11  TARGET(PRICE_IN_LACS)  29451 non-null  float64
dtypes: float64(4), int64(5), object(3)
memory usage: 2.7+ MB


None

- The train dataset contains **29,451** instances with **12** fratures recorded for each instance.
- There are three features of the `object` type which are `categorical` in the case of a pandas dataframe. These columns are `POSTED_BY`, `BHK_OR_RK` and `ADDRESS`.
- The data contains four `float` features: The area covered by the house recorded as `SQUARE_FT`, latitude, longitude, (recorded as `LATITUDE`, `LONGITUDE` respectively) and the target feature **price** recorded as `TARGET(PRICE_IN_LACS)`.
- The dataset contains five `int` features: `UNDER_CONSTRUCTION`, `RERA`, `BHK_NO`, `BHK_OR_RK`, `READY_TO_MOVE` and `RESALE`. Integers are whole numbers but often, datatypes recorded as integers store binary (0, 1) values. In other cases, it contains ordinal data, the cases will be processed accordingly.

### **Actions to be taken:**
- The inconsistencies in the column names will be handled in the **data cleaning** phase of the project. For instance, the target column will be remaned to `price` and all column names set to lower case and renamed to their respective meaning.
- Take note of the columns (`int`) that stored binary values.
- Take note for the object columns, and think about insights to extract from them and the right tools to use. 

In [7]:
inspector.missing_values();


No missing values found.


- The dataset contains **0** missing values.

In [8]:
inspector.duplicates();


Number of duplicate rows: 401


The dataset contains **401** duplicate rows. These will be handled in the data cleaning phase of the project.

## **Summary Statistics**

In [9]:
inspector.statistics();


Summary Statistics: (including categorical data)
------------------------------


POSTED_BY  UNDER_CONSTRUCTION          RERA       BHK_NO. BHK_OR_RK  \
count      29451        29451.000000  29451.000000  29451.000000     29451   
unique         3                 NaN           NaN           NaN         2   
top       Dealer                 NaN           NaN           NaN       BHK   
freq       18291                 NaN           NaN           NaN     29427   
mean         NaN            0.179756      0.317918      2.392279       NaN   
std          NaN            0.383991      0.465675      0.879091       NaN   
min          NaN            0.000000      0.000000      1.000000       NaN   
25%          NaN            0.000000      0.000000      2.000000       NaN   
50%          NaN            0.000000      0.000000      2.000000       NaN   
75%          NaN            0.000000      1.000000      3.000000       NaN   
max          NaN            1.000000      1.000000     20.000000       NaN   

           SQUARE_FT  READY_TO_MOVE        RESALE              ADDRESS  \
count   2.945100e+04   29451.000000  29451.000000                29451   
unique           NaN            NaN           NaN                 6899   
top              NaN            NaN           NaN  Zirakpur,Chandigarh   
freq             NaN            NaN           NaN                  509   
mean    1.980217e+04       0.820244      0.929578                  NaN   
std     1.901335e+06       0.383991      0.255861                  NaN   
min     3.000000e+00       0.000000      0.000000                  NaN   
25%     9.000211e+02       1.000000      1.000000                  NaN   
50%     1.175057e+03       1.000000      1.000000                  NaN   
75%     1.550688e+03       1.000000      1.000000                  NaN   
max     2.545455e+08       1.000000      1.000000                  NaN   

           LONGITUDE      LATITUDE  TARGET(PRICE_IN_LACS)  
count   29451.000000  29451.000000           29451.000000  
unique           NaN           NaN                    NaN  
top              NaN           NaN                    NaN  
freq             NaN           NaN                    NaN  
mean       21.300255     76.837695             142.898746  
std         6.205306     10.557747             656.880713  
min       -37.713008   -121.761248               0.250000  
25%        18.452663     73.798100              38.000000  
50%        20.750000     77.324137              62.000000  
75%        26.900926     77.828740             100.000000  
max        59.912884    152.962676           30000.000000


Unique Values per Categorical Column:
------------------------------

Column: POSTED_BY (Unique Values: 3)
------------------------------


array(['Owner', 'Dealer', 'Builder'], dtype=object)


Column: UNDER_CONSTRUCTION (Unique Values: 2)
------------------------------


array([0, 1])


Column: RERA (Unique Values: 2)
------------------------------


array([0, 1])

There are more than 10 unique values.

Column: BHK_OR_RK (Unique Values: 2)
------------------------------


array(['BHK', 'RK'], dtype=object)


Column: READY_TO_MOVE (Unique Values: 2)
------------------------------


array([1, 0])


Column: RESALE (Unique Values: 2)
------------------------------


array([1, 0])

There are more than 10 unique values.


- All features contains the same number of instances (**29,451**). This is possible because the dataset contains no missing values, as seen in the inspection in the section above.
- `POSTED_BY` has three unique values: **Owner**, **Dealer**, **Builder**, with **Dealer** being the most occuring value (mode). This is reasonable as dealers are professionals at marketing real-estate products. Out of the **29,451** houses, **18,291** are posted by the dealer.
- `UNDER_CONSTRUCTION` has two unique values (0, 1), meaning this is a binary colum  indicating if the house is **under construction** or **not**. This feature is an `integer` feature with a mean of **0.179756** with a standard deviation of **0.383991**. The value of the mean is closer to 0 which suggests that most houses are **not under construction** (more **0** values for this feature). This is also confirmed by the first, second, and third quartile which are all zeros. The data ranges between 0 and 1, and with a standard deviation of **0.383991**, a three-standard deviation ($3 \times 0.383991 = 1.153$) which is only a bit above the maximum value in the feature distribution.
- f  

In [10]:
3 * 0.383991

1.1519730000000001